### PR TITLE
mm/kmm_map: Add support to dynamically map pages into kernel virtual …

### PR DESCRIPTION
**Pull request by pussuw**
…memory

## Summary
This adds functionality to map pages dynamically into kernel virtual memory. This allows implementing I/O remap for example, which is a useful (future) feature.

Now, the first target is to support mapping user pages for the kernel.

Why? There are some userspace structures that might be needed when the userspace process is not running. Semaphores are one such example. Signals and the WDT timeout both need access to the user semaphore to work properly. Even though for this only obtaining the kernel addressable page pool virtual address is needed, for completeness a procedure is provided to map several pages.

NOTE: ARM target does not implement support yet, I'm not sure how to implement it.
## Impact
Add new feature for CONFIG_BUILD_KERNEL, nothing else is affected
## Testing
icicle:knsh

### CODE REVIEW

**Comment by patacongo** [comment]
> @pussuw is there some test that could be add to ostest to verify if dynamic map pages is working? Also please consider added some basic Documentation/ it will help people discover that NuttX supports it.

ostest does not currently work in KERNEL mode.  It relies things like task_create() which don't exist in KERNEL mode.

**Comment by pussuw** [comment]
@acassis like @patacongo said currently ostest does not work at all for kernel mode. I tried doing an intermediate step https://github.com/apache/nuttx-apps/pull/1500 by forwarding task_create to pthread_create but that is not the same thing. The correct implementation would be to go through the tests case-by-case and:
- When task_create is used to make workers, they can be replaced with pthreads
- When task_create is used to make other "processes", they must be replaced with posix_spawn() or similar (as we don't have fork()). Making the tests compatible to be spawned requires a bit of refactoring.

As a summary; making all ostest steps work in kernel mode is doable but not an insignificant amount of work which I simply do not have time to do now.

File: include/nuttx/arch.h [hunk 1]
```diff
@@ -1253,6 +1253,108 @@ int up_addrenv_kstackalloc(FAR struct tcb_s *tcb);
...
+ *   True if it is; false if it's not
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_ARCH_ADDRENV
+bool up_addrenv_user_vaddr(uintptr_t vaddr);
+#endif
+
+/****************************************************************************
+ * Name: riscv_map_pages
```
Comment (pussuw) [inline include/nuttx/arch.h:1313]: Note to self: copy&paste error, both here and in implementation

**Comment by patacongo** [comment]
> * When task_create is used to make workers, they can be replaced with pthreads

That most likely works because NuttX uses heavyweight threads, ie., coupled tasks that masquerade as pthreads.  So IPCs like semaphores, legacy signal functions, etc. work okay because PIDs, TIDs, and pthread_t's are are aliased; semaphore shared attributes are ignored; etc.  Certainly the result would not be POSIX compliant and so not a proper solution for a POSIX compliance test.  In order to be a POSIX compliant test, the test it would have to use pthread_sigprocmask() vs. sigprocmask(), for examples.  Signals are dispatched completely differently for pthread_kill() and kill().

It might make more sense to create a separate test for unique kernel build features.  Or maybe two versions of OS test that share as much they can.

Or maybe force the existing OS test execute with a file system and all tasks as files in the file system.  Then perhaps they could be made compatible.

**Comment by pussuw** [comment]
> Certainly the result would not be POSIX compliant and so not a proper solution for a POSIX compliance test. In order to be a POSIX compliant test, the test it would have to use pthread_sigprocmask() vs. sigprocmask(), for examples. Signals are dispatched completely differently for pthread_kill() and kill().

Yes indeed, using pthreads was a kludge solution (or maybe even a dirty hack) and that is why the PR is pending.

> Or maybe two versions of OS test that share as much they can.

I think this is unavoidable, some split is expected.

> Or maybe force the existing OS test execute with a file system and all tasks as files in the file system. Then perhaps they could be made compatible.

I think kernel mode needs this anyway, due to not having fork() i.e. only spawning from file works. I had some sort of idea to keep the current code pretty much compatible with the flat addressing modes too with wrapping / flagging. Your solution might be better but it does force compiling the tests into loadable files -> need more memory and features enabled to test the os.

**Comment by patacongo** [comment]
> I think this is unavoidable, some split is expected.

The basic tests should be all/mostly re-usable.  It really should be a big but low risk packaging change.

> I think kernel mode needs this anyway, due to not having fork() i.e. only spawning from file works.

Does vfork not work?  fork() or vfork() would ultimately be followed by exec() and so would still need to be exec'ed from a file.

This is a good discussion, but unfortunately off-topic for this PR.

**Comment by pussuw** [comment]
>The basic tests should be all/mostly re-usable. It really should be a big but low risk packaging change.
Yes agree, mostly the signal/waitpid and such tests need to be refactored.

> Does vfork not work? fork() or vfork() would ultimately be followed by exec() and so would still need to be exec'ed from a file.
I was thinking task_spawn in this case is sort of just fork()'ing without exec(). vfork IMHO is superseded by posix_spawn anyway and both need a file to exec(). This should all be doable with posix_spawn and I in fact started writing a wrapper for task_create that uses posix_spawn() but then something more important came up.

> This is a good discussion, but unfortunately off-topic for this PR.

I do agree with Alan, we should have a test suite for these new features. However as AFAIK there are not many users of kernel mode right now / yet, and I can personally just use my application (which uses a wide POSIX feature set) to test the features I need. So again, writing the kernel mode support for ostest is unfortunately very low priority for me right now :(

**Comment by patacongo** [comment]
> I was thinking task_spawn in this case is sort of just fork()'ing without exec().

task_spawn() is not available in KERNEL mode either:

    138 #ifndef CONFIG_BUILD_KERNEL
    139 /* Non-standard task_spawn interface.  This function uses the same
    140  * semantics to execute a file in memory at 'entry', giving it the name
    141  * 'name'.
    142  */
    143
    144 int task_spawn(FAR const char *name, main_t entry,
    145       FAR const posix_spawn_file_actions_t *file_actions,
    146       FAR const posix_spawnattr_t *attr,
    147       FAR char * const argv[], FAR char * const envp[]);
    148 #endif

Nothing that includes the address of main() can be used to start tasks in that mode since the address of main() is not available in that address environment.  You would have to use posix_spawn() which does take a file path instead of an address.  posix_spawn() is essentially a fork+exec.

**Comment by pussuw** [comment]
What I mean is with fork() you could do something like overload the implementation of task_spawn (in ostest) like this:

```
int task_spawn(FAR const char *name, main_t entry,
       FAR const posix_spawn_file_actions_t *file_actions, /* This can be ignored afaik ? */
       FAR const posix_spawnattr_t *attr, /* This can also be ignored afaik */
       FAR char * const argv[], FAR char * const envp[])
{
  int pid = fork();
  if (pid == 0)
    {
      /* Child performs the task given to task_spawn() */
      
      entry(argc, argv);
    }
    
  /* Parent returns the child pid to caller */
  
  return pid;
}
```
Unless I'm missing something this should work (might not do what we want it to do, but works), as the main thread does indeed exist in the same address environment. Maybe there is something else I'm missing.

Using posix_spawn is the best option still, just need to make at least some of the tests into loadable processes and that should do it.

**Review by patacongo — CHANGES_REQUESTED** [review]
Most of my review was kind of superficial because I am not familiar with RISC-V.  However, it appears to me that you have implement a parallel implementation to the existing, established, documented page allocator design prototype typed in include/nuttx/arch.h:

     671 /****************************************************************************
     672  * Name: pgalloc
     673  *
     674  * Description:
     675  *   If there is a page allocator in the configuration and if an MMU is
     676  *   available to map physical addresses to virtual address, then this
     677  *   function must be provided by the platform-specific code.  This is part
     678  *   of the implementation of sbrk().  This function will allocate the
     679  *   requested number of pages using the page allocator and map them into
     680  *   consecutive virtual addresses beginning with 'brkaddr'
     681  *
     682  *   NOTE:  This function does not use the up_ naming standard because it
     683  *   is indirectly callable from user-space code via a system trap.
     684  *   Therefore, it is a system interface and follows a different naming
     685  *   convention.
     686  *
     687  * Input Parameters:
     688  *   brkaddr - The heap break address.  The next page will be allocated and
     689  *     mapped to this address.  Must be page aligned.  If the memory manager
     690  *     has not yet been initialized and this is the first block requested for
     691  *     the heap, then brkaddr should be zero.  pgalloc will then assigned the
     692  *     well-known virtual address of the beginning of the heap.
     693  *   npages - The number of pages to allocate and map.  Mapping of pages
     694  *     will be contiguous beginning beginning at 'brkaddr'
     695  *
     696  * Returned Value:
     697  *   The (virtual) base address of the mapped page will be returned on
     698  *   success.
     699  *   Normally this will be the same as the 'brkaddr' input. However, if
     700  *   the 'brkaddr' input was zero, this will be the virtual address of the
     701  *   beginning of the heap.  Zero is returned on any failure.
     702  *
     703  ****************************************************************************/
     704
     705 #if defined(CONFIG_ARCH_ADDRENV) && defined(CONFIG_MM_PGALLOC) && \
     706     defined(CONFIG_ARCH_USE_MMU)
     707 uintptr_t pgalloc(uintptr_t brkaddr, unsigned int npages);
     708 #endif

That is the interface currently used by Used by sbrk, initialization logic, shm interfaces.

I don't think we should permit multiple different designs for essentially the same thing in the OS.  That would be the wrong way to go.  So I think we would either have to reject this implementation OR unify the two implementations into one standard page allocator interface that is used throughout the OS and enforce that common interface in all future architectures.

Fortunately, that is not much in place in the repository now.  Only two ARMv7-A archtectures other than RISC-V claim KERNEL build support now :

    $ find . -name defconfig | xargs grep BUILD_KERNEL
    nuttx/arch/arm/imx6/sabre-6quad/configs/knsh/defconfig:CONFIG_BUILD_KERNEL=y
    nuttx/arch/arm/imx6/sabre-6quad/configs/knsh_smp/defconfig:CONFIG_BUILD_KERNEL=y
    nuttx/arch/arm/imx6/sabre-6quad/configs/netknsh/defconfig:CONFIG_BUILD_KERNEL=y
    nuttx/arch/arm/imx6/sabre-6quad/configs/netknsh_smp/defconfig:CONFIG_BUILD_KERNEL=y
    nuttx/arch/arm/sama5/sama5d4-ek/configs/knsh/defconfig:CONFIG_BUILD_KERNEL=y

For both architectures, the page allocator support is in:

    arm/src/armv7-a/arm_pgalloc.c
    nuttx/mm/mm_gran/mm_pgalloc.c

We can't let these architectures diverge to the degree that merging this change would cause.  We need to spend more time making the solution into one, common, standard OS architecture.  That is very important.

**Comment by xiaoxiang781216** [comment]
> > task_spawn() is not available in KERNEL mode either:
> 
> My bad, I meant task_create of course.. Has been a long day...
> 
> What I mean is with fork() you could do something like overload the implementation of task_create (in ostest) like this:
> 
> ```
> int task_create(FAR const char *name, int priority,
>                 int stack_size, main_t entry, FAR char * const argv[])
> {
>   int pid = fork();
>   if (pid == 0)
>     {
>       /* Child performs the task given to task_spawn() */
>       ... calculate argc etc ...
>       entry(argc, argv);
>     }
>     
>   /* Parent returns the child pid to caller */
>   
>   return pid;
> }
> ```
> 
> Unless I'm missing something this should work (might not do what we want it to do, but works), as the main thread does indeed exist in the same address environment. Maybe there is something else I'm missing but as far as I can tell task_create() is in fact quite close to fork() (without the obvious address environment stuff).
> 
> Using posix_spawn is the best option still, just need to make at least some of the tests into loadable processes and that should do it.
> 

We can keep the current ostest structure even switch to posix_spawn:

1. Assign the test case launched by task_create an unique id(string is a good choice)
2. If ostest is launched with one argument, do the test identified by the unique id
3. Change task_create to posix_spawn with "ostest" and the unique id

posix_spawn work on all mode(FLAT, PROTECTED and KERNEL), so we can use the same code base.

File: arch/risc-v/src/common/riscv_mmu.h [hunk 2]
```diff
@@ -308,24 +308,40 @@ static inline uintptr_t mmu_pte_to_paddr(uintptr_t pte)
...
+ *   Physical address from SATP value
  *
  ****************************************************************************/
 
-static inline uintptr_t mmu_get_satp_pgbase(void)
+static inline uintptr_t mmu_satp_to_paddr(uintptr_t satp)
 {
   uintptr_t ppn;
-  ppn = mmu_read_satp();
+  ppn = satp;
```
[the remark this answers is not in the document]
Comment (pussuw) [reply arch/risc-v/src/common/riscv_mmu.h:324]: Yes making a copy is not -strictly- necessary here. I presume the compiler sees this and does not allocate separate storage for ppn anyway.

**Comment by pussuw** [comment]
>For both architectures, the page allocator support is in:

>```
arm/src/armv7-a/arm_pgalloc.c
nuttx/mm/mm_gran/mm_pgalloc.c
```

I implemented the riscv address environment support long ago, it implements exactly the same public API:
https://github.com/apache/nuttx/blob/master/arch/risc-v/src/common/riscv_pgalloc.c

File: arch/risc-v/src/common/addrenv.h [hunk 3]
```diff
@@ -44,6 +44,14 @@
...
 #endif
 
+/* Base address for address environment */
+
+#if CONFIG_ARCH_TEXT_VBASE != 0
+#  define ARCH_ADDRENV_VBASE    (CONFIG_ARCH_TEXT_VBASE)
+#else
+#  define ARCH_ADDRENV_VBASE    (CONFIG_ARCH_DATA_VBASE)
+#endif
+
```
[the remark this answers is not in the document]
Comment (pussuw) [reply arch/risc-v/src/common/addrenv.h:54]: Done

File: arch/risc-v/src/common/pgalloc.h [hunk 4]
```diff
@@ -78,6 +80,24 @@ static inline uintptr_t riscv_pgvaddr(uintptr_t paddr)
...
+ ****************************************************************************/
+
+static inline bool riscv_uservaddr(uintptr_t vaddr)
+{
+  /* Check if this address is within the range of the virtualized .bss/.data,
+   * heap, or stack regions.
+   */
+
+  return vaddr >= ARCH_ADDRENV_VBASE;
+}
```
[the remark this answers is not in the document]
Comment (pussuw) [reply arch/risc-v/src/common/pgalloc.h:99]: Done

**Review by patacongo — APPROVED** [review]

File: mm/kmap/kmm_map.c [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+
+  /* Attempt to map the pages */
+
+  vaddr = (uintptr_t)map_pages(pages, npages, prot);
+  if (!vaddr)
+    {
+      return NULL;
+    }
+
+  return (FAR void *)(vaddr);
```
Comment (xiaoxiang781216) [inline mm/kmap/kmm_map.c:281]: remove () around vaddr

File: arch/risc-v/src/common/riscv_addrenv_pgmap.c [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+ * Input Parameters:
+ *   addrenv - The user address environment.
+ *   vaddr   - The user virtual address
+ *
+ * Returned Value:
+ *   Page physical address on success; NULL on failure.
+ *
+ ****************************************************************************/
+
+uintptr_t up_addrenv_find_page(FAR arch_addrenv_t *addrenv, uintptr_t vaddr)
```
Comment (pkarashchenko) [inline arch/risc-v/src/common/riscv_addrenv_pgmap.c:62]: Remove FAR

File: include/nuttx/mm/kmap.h [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __INCLUDE_NUTTX_MM_KMAP_H_
```
Comment (pkarashchenko) [inline include/nuttx/mm/kmap.h:21]: Remove `_` after `H`

File: mm/kmap/kmm_map.c [hunk 8]
```diff
@@ -0,0 +1,466 @@
...
+ *   pages  - Pointer to buffer that contains the physical page addresses.
+ *   npages - Amount of pages.
+ *   prot   - Access right flags.
+ *
+ * Returned Value:
+ *   Pointer to the mapped virtual memory on success; NULL on failure
+ *
+ ****************************************************************************/
+
+static void *map_pages(FAR void **pages, size_t npages, int prot)
```
Comment (pkarashchenko) [inline mm/kmap/kmm_map.c:115]: `static FAR void *`

File: mm/kmap/kmm_map.c [hunk 9]
```diff
@@ -0,0 +1,466 @@
...
+ *   pages  - Pointer to buffer that contains the physical page addresses.
+ *   npages - Amount of pages.
+ *   prot   - Access right flags.
+ *
+ * Returned Value:
+ *   Pointer to the mapped virtual memory on success; NULL on failure
+ *
+ ****************************************************************************/
+
+static void *map_single_user_page(uintptr_t vaddr)
```
Comment (pkarashchenko) [inline mm/kmap/kmm_map.c:178]: `static FAR void *`

File: mm/kmap/kmm_map.c [hunk 10]
```diff
@@ -0,0 +1,466 @@
...
+    {
+      /* Yes, can simply return the kernel addressable virtual address */
+
+      vaddr = (uintptr_t)map_single_user_page(vaddr);
+      return (FAR void *)(vaddr + offset);
+    }
+
+  /* No, the area must be mapped into kernel virtual address space */
+
+  pages = kmm_zalloc(npages * sizeof(void *));
```
Comment (pkarashchenko) [inline mm/kmap/kmm_map.c:388]: `sizeof(FAR void *)`

Comment (pussuw) [reply mm/kmap/kmm_map.c:388]: Unnecessary, the size of void* does not change if declared far

Comment (pkarashchenko) [reply mm/kmap/kmm_map.c:388]: Just for style consistency.

Comment (pkarashchenko) [reply mm/kmap/kmm_map.c:388]: Yeah, just rechecked. 16 bits for near and 32 bits for far

**Pull request merged**